### PR TITLE
Add Docker DOOD setup

### PR DESCRIPTION
**Pull request by rbren**
This builds off of work from several people:
* @leehanchung
* @spuder
* @foragerr 
* @pranjallk1995

And probably more I'm forgetting.

The biggest change here is in workspace directory management. This is kind of broken today--there are places that use `config.WORKSPACE_DIR`, while others respect local settings. We're not consistent about this.

I created `WORKSPACE_BASE`, which is a global setting. There's also `WORKSPACE_MOUNT_PATH`, which controls the directory passed to `-v` when starting the sandbox (it has to reference the host machine in a DOOD setup).

In the future, I think we'll want to have a `directory` argument for the agent/sandbox again, which will specify a _subdirectory_ of `WORKSPACE_BASE`. That way the mount for the app container stays the same, and we just adjust the mount for the sandbox. And we'll be able to run multiple agents in different directories concurrently.

TODO:
* [x] Remove need for ExecBox
* [x] Better workspace mounting
* [x] More efficient build
* [x] Compile frontend assets instead of using dev server
* [x] Makefile for pushing

### Test Cases
For each of these cases, we need to make sure:
* The agent can write/read a file
* The agent can run the file
* The file persists in the specified workspace after the agent is done

#### Normal one-off run
```bash
poetry run python opendevin/main.py -d ./workspace/ -t "write a bash script that reverses a string to a file called rev.sh and test it"
```

#### ExecBox one-off run
```bash
SANDBOX_TYPE=exec poetry run python opendevin/main.py -d ./workspace/ -t "write a bash script that reverses a string to a file called rev.sh and test it"
```

#### Docker build
```bash
docker build -t opendevin-app -f container/Dockerfile .
```

#### Docker one-off run
```bash
export WORKSPACE_DIR=$(pwd)/workspace

docker run \
    -e LLM_API_KEY \
    -e WORKSPACE_MOUNT_PATH=$WORKSPACE_DIR \
    -v $WORKSPACE_DIR:/opt/workspace_base \
    -v /var/run/docker.sock:/var/run/docker.sock \
    opendevin-app \
    poetry run python opendevin/main.py \
    -t "write a bash script that reverses a string to a file called rev.sh and test it"
```

#### Docker app
```bash
export WORKSPACE_DIR=$(pwd)/workspace

docker run \
    -e LLM_API_KEY \
    -e WORKSPACE_MOUNT_PATH=$WORKSPACE_DIR \
    -v $WORKSPACE_DIR:/opt/workspace_base \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -p 2000:3000 \
    opendevin-app
```

### CODE REVIEW

**Comment by foragerr** [comment]
This looks very close to what I had done in #110 ; I also completed:
* replace execbox with "localbox"
* compile front end, and stick it on a fastapi path so; docker only runs 1 active process

But I don't fully understand the jwt issue you ran into - I'm unable to repro it.
If your setup works, let's go with that.

I'm going a raise a separate PR to add just localbox, so you can use it with your setup if you like.

**Comment by spuder** [comment]
This does look relatively clean.

One criticism is running 2 processes inside 1 container. Best practice is to have 1 process per container because docker is effectively replacing the init system and it tracks the status of pid 1.

By running multiple processes you loose the ability to track if a single process dies. Nor can you send term signals. In a development environment it probably doesn't matter much.

**Comment by rbren** [comment]
@foragerr do you have a branch/PR where I can see your work? I'm happy to take a look at it

@spuder I agree--we will definitely want to split BE and FE docker images at some point. I'm torn on whether this should be our "quickstart" image, or if we should use `docker compose`

**Comment by foragerr** [comment]
https://github.com/foragerr/OpenDevin/pull/10/files#diff-b09b9fb99ae1ede0edf0fe18e34777e67bb38da89272f7af53db8880794437a4

This multistage dockerfile compiles front end and adds it to fastAPI as a static site.
fastAPI listener paths need a bit of adjustment, also in that PR.

**Comment by rbren** [comment]
OK awesome, there's some great stuff in here, like mounting on `/api`

**Comment by spuder** [comment]
> I'm torn on whether this should be our "quickstart" image, or if we should use docker compose

If the goal is just to containerize the app so that people don't have to mess with python environments dependencies, then either is fine. 

If the goal is to provide a standard framework for people to build up the entire stack in a repeatable way, then I would lean towards compose. 

One caveat of compose is that while it provides a consistent orchestration framework, there is some tech debt between `docker compose` and `docker-compose` that often trips people up. Detailed explanation of the [quirks here](https://github.com/Ankermgmt/ankermake-m5-protocol/issues/116#issuecomment-1563186853)

**Comment by rbren** [comment]
Oh funny, I was wondering what the difference between those two was!

Near-term, the goal is definitely to just get rid of all the frustrating python dependency issues that keep flooding the repo.

Long-term, I definitely want a scalable/repeatable stack, with separated FE and BE containers

**Comment by foragerr** [comment]
@spuder this is my thinking on that exact topic (poster earlier in #110):

> Sounds like you'll want to eventually support multiple deployment modes -
> 1. If you're an OpenDevin developer; do something very close to what's available today, Makefiles, poetry, pnpm etc
> 2. If you're spinning up a multi-tenant production setup, you may get separate backend, frontend images, perhaps externalized vector store, perhaps E2B for sandboxes
> 3. If you're just looking to kick the tires on OpenDevin the simplest/fastest not-prodution-safe way, run this docker run command.
> I'm going to focus on creating a proof-of-concept for # 3

**Comment by rbren** [comment]
OK I made some changes based on the feedback:
* Multi-arch build, cribbing from @foragerr 
* Frontend is built statically, so now just one binary runs (also stole @foragerr's code here 😄)
* SSHBox is now used
  * had to do some work on networking and userspace to make this happen

Only thing I'd really like to do now is automate the build and publish process

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+
+FROM python:3.12-slim as runtime
+
+WORKDIR /app
+ENV PYTHONPATH '/app'
+
+RUN apt-get update -y
+RUN apt-get install -y curl make git python3.11 python3-pip
+RUN curl -fsSL https://get.docker.com | sh
```
Comment (foragerr) [inline Dockerfile:19]: FWIW, I didn't need to install docker. I didn't poke at it a lot, but OpenDevin server seemed to be able to make a launch a sandbox even if I didn't explicitly intall docker on python:slim. So I skipped it.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+RUN npm run build
+
+
+FROM python:3.12-slim as runtime
+
+WORKDIR /app
+ENV PYTHONPATH '/app'
+
+RUN apt-get update -y
+RUN apt-get install -y curl make git python3.11 python3-pip
```
Comment (foragerr) [inline Dockerfile:18]: is there a reason to install 3.11 in a 3.12 image?
the runtime image should also already have pip.

**Comment by rbren** [comment]
OK this is working now! A bit about volume mounting:

When we create a docker container from _inside_ docker, but using DOOD, the mount path has to be _on the host machine_. That makes it really tricky to coordinate the workspace.

So, if the workspace is, say, `~/foo`
* the user mounts `~/foo` to `/workspace` in the app container
* the app container mounts `~/foo` to `/workspace` in the sandbox

In the future, I think what we'll want to do is mount to a base dir, where the user stores all their projects. Then the UI can show a home-grown file explorer, where they can pick which directory they want to work in today.

File: Dockerfile (hunk 2 above)
Comment (rbren) [reply Dockerfile:18]: Whoops good catch, copy/pasta

File: Dockerfile (hunk 1 above)
Comment (rbren) [reply Dockerfile:19]: Hmm--I'll give it a shot

Comment (spuder) [reply Dockerfile]: Same, I didn't need the docker binary installed on my branch, the python library just needs the socket.

**Comment by spuder** [comment]
> When we create a docker container from inside docker, but using DOOD, the mount path has to be on the host machine. That makes it really tricky to coordinate the workspace.

I also ran into that. 

As an aside, It becomes extra tricky for podman users as the podman machine has selinux enabled by default and also has many paths read only (/usr being one of them). (My work policy does not allow docker desktop to be installed, but does allow podman as it is part of the RedHat ecosystem).

While I don't expect podman to be supported for openDevin, I would put a vote for using `/opt` or `~/` paths as they are less locked down than `/` and `/usr`

**Comment by loopyd** [comment]
Just a side project, but if any code is useful here, happy to contribute it.  I was inspired by @pranjallk1995 with mine when I'd seen their original commit come in.  I figured I may as well contribute my full ``docker-entrypiont.sh``, and some of my bash stack/swarm API code, as well as my command line utility that manages the docker stack and stack compose files.  I've been pumping quite a lot of time into mine, so have at, current contributors.

[DockerOpenDevin on GitHub](https://github.com/loopyd/DockerOpenDevin)

**Comment by rbren** [comment]
@loopyd thanks for all that! It's super helpful.

I would love to add a DIND solution to the mix here--let's dig into your work once the DOOD solution is in

**Comment by rbren** [comment]
This is ready for final review! Everything seems to be working well, including local dev. 

The core issue was needing to rework workspace directory management. It was actually kind of broken as-is. Some notes in the description

**Comment by foragerr** [comment]
@rbren are you able to publish an image using the Dockerfile in your PR?

File: container/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+WORKDIR /app
+ENV PYTHONPATH '/app'
+ENV RUN_AS_DEVIN=false
+ENV USE_HOST_NETWORK=false
+ENV SSH_HOSTNAME=host.docker.internal
+ENV WORKSPACE_BASE=/opt/workspace_base
+RUN mkdir -p $WORKSPACE_BASE
+
+RUN apt-get update -y
+RUN apt-get install -y curl make git build-essential
```
Comment (foragerr) [inline container/Dockerfile:23]: `build-essential` is a pretty large installation, with lots of dependencies that get installed - and does not need to be retained in the runtime image. It's the whole reason I added a separate python-build stage here: https://github.com/foragerr/OpenDevin/pull/10/files#diff-b09b9fb99ae1ede0edf0fe18e34777e67bb38da89272f7af53db8880794437a4R10-R26

File: container/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+RUN python3 -m pip install poetry  --break-system-packages
+
+COPY ./pyproject.toml ./poetry.lock ./
+RUN poetry install --without evaluation
+
+COPY ./opendevin ./opendevin
+COPY ./agenthub ./agenthub
+RUN poetry run python opendevin/download.py # No-op to download assets
```
Comment (foragerr) [inline container/Dockerfile:31]: Consider combining all the RUN commands for fewer layers.
perhaps:


```
COPY ./pyproject.toml ./poetry.lock ./
COPY ./opendevin ./opendevin
COPY ./agenthub ./agenthub

RUN mkdir -p $WORKSPACE_BASE && \
    apt-get update -y && \
    apt-get install -y curl make git build-essential && \
    python3 -m pip install poetry  --break-system-packages && \
    poetry install --without evaluation && \
    poetry run python opendevin/download.py # No-op to download assets
```

File: container/Dockerfile (hunk 3 above)
Comment (foragerr) [inline container/Dockerfile:23]: apt update and install should go in the same RUN statement. 
https://docs.docker.com/develop/develop-images/instructions/#apt-get

File: opendevin/controller/command_manager.py [hunk 5]
```diff
@@ -8,26 +8,23 @@
...
-        self.directory = directory
         sandbox_type = config.get(ConfigType.SANDBOX_TYPE).lower()
         if sandbox_type == 'exec':
             self.shell = DockerExecBox(
-                sid=(sid or 'default'), workspace_dir=directory, container_image=container_image
+                sid=(sid or 'default'), container_image=container_image
             )
         elif sandbox_type == 'local':
-            self.shell = LocalBox(workspace_dir=directory)
+            self.shell = LocalBox()
```
Comment (foragerr) [inline opendevin/controller/command_manager.py:24]: Does LocalBox need a change to deal with not being given a workspace directory?

**Comment by foragerr** [comment]
Woah, that's a lot of change!
Thanks for doing this @rbren

**Comment by foragerr** [comment]
We probably need an Actions workflow to publish this image on merge to main. (Doesn't need to be in this PR)
How is the sandbox image published right now?

**Comment by foragerr** [comment]
@enyst did you mean port `3000` ?

**Comment by loopyd** [comment]
> Woah, that's a lot of change! Thanks for doing this @rbren

Too right, he's definitely being smart about it and consolidating the work.  Good job @rbren

**Comment by enyst** [comment]
> @enyst did you mean port `3000` ?

In the README, we say to run `docker run ... -p 2000:3000`, isn't that 2000 on the host? @foragerr

**Comment by foragerr** [comment]
You're absolutely right, sorry I missed that.
Although, I would prefer it to stay 3000 across the board, that's how we had always done it.

File: container/Dockerfile (hunk 4 above)
Comment (rbren) [reply container/Dockerfile:31]: I did some combining here, but I generally find it's better to put dependencies in one layer, then source code in a later later. That way, as you change the source code, you can continue using the cached dependency layer.

File: container/Makefile [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+MAJOR_FULL_IMAGE=$(DOCKER_BUILD_REGISTRY)/$(DOCKER_BUILD_ORG)/$(DOCKER_BUILD_REPO):$(MAJOR_VERSION)
+MINOR_VERSION=$(shell echo $(DOCKER_BUILD_TAG) | cut -d. -f1,2)
+MINOR_FULL_IMAGE=$(DOCKER_BUILD_REGISTRY)/$(DOCKER_BUILD_ORG)/$(DOCKER_BUILD_REPO):$(MINOR_VERSION)
+
+# normally, for local build testing or development. use cross platform build for sharing images to others.
+build:
+	docker build -f Dockerfile -t ${FULL_IMAGE} -t ${LATEST_FULL_IMAGE} ..
+
+push:
+	docker push ${FULL_IMAGE} ${LATEST_FULL_IMAGE}
```
[the remark this answers is not in the document]
Comment (rbren) [reply container/Makefile:19]: I think we get that for free from GitHub actions--at least that's how the existing images seem to work

File: container/Dockerfile (hunk 3 above)
Comment (rbren) [reply container/Dockerfile]: Good call!

Comment (rbren) [reply container/Dockerfile]: Good point! I was actually wondering about the python build stage.

Mind if we tackle this in a follow-on?

File: opendevin/controller/command_manager.py (hunk 5 above)
Comment (rbren) [reply opendevin/controller/command_manager.py:24]: Yup! Good catch.

Do you have a good (non-scary) way to test locally? Maybe we should add a `__main__` to LocalBox so it's easier to test

**Comment by rbren** [comment]
> We probably need an Actions workflow to publish this image on merge to main. (Doesn't need to be in this PR)
How is the sandbox image published right now?

I think this is all handled by our GitHub action automatically, CC @iFurySt 

https://github.com/OpenDevin/OpenDevin/blob/main/.github/workflows/ghcr.yml

Re: port 3000, totally agree, fixed.

Image is now available as `ghcr.io/rbren/opendevin`!

**Comment by foragerr** [comment]
You may need to make that image repo public

File: opendevin/controller/command_manager.py (hunk 5 above)
Comment (foragerr) [reply opendevin/controller/command_manager.py:24]: I tend to test inside docker. I'll take a look at this after the PR merges. It can wait.

**Review by foragerr — APPROVED** [review]

**Comment by rbren** [comment]
Whoops public now. Will leave this open a bit longer for testing/comments

File: frontend/src/types/ConfigType.tsx [hunk 7]
```diff
@@ -19,7 +19,7 @@ enum ArgConfigType {
 const SupportedList: string[] = [
   // ArgConfigType.LLM_API_KEY,
   // ArgConfigType.LLM_BASE_URL,
-  ArgConfigType.WORKSPACE_DIR,
+  // ArgConfigType.WORKSPACE_DIR,
```
Comment (808vita) [inline frontend/src/types/ConfigType.tsx:22]: so option to set workspace from the frontend is temporarily removed

Comment (rbren) [reply frontend/src/types/ConfigType.tsx:22]: Correct--it's actually broken right now. We'll re-add it in a future PR

**Comment by enyst** [comment]
This is so awesome!

**Pull request merged**